scala/toolchains/README.md

```diff
@@ -0,0 +1,153 @@
+# Developing Toolchains for Deps
```

Comment:
@katre is there a chance you can take a look at this toolchains pattern?
I think this might be valuable for rulesets in general and not only rules_scala.
If you think we're abusing toolchains or just misunderstanding it we'd also love to know that.
One concrete question is whether using a map [here](https://github.com/bazelbuild/rules_scala/pull/1072/files#diff-4315eb5f77e2ca3a6bb2a3af7f783366R19) is the best solution?

Reply:
I looked over the PR and the docs, and I'm afraid I don't fully understand what's being implemented. You are adding dependencies to the toolchain that will then also be provided to other, non-toolchain-aware rules, via the normal transitive dependencies? Can you give a concrete example of where this would be useful?

As for the map, I am also confused about why it is a map from a label (which seems to always be a declare_deps_provider) to a string (which is a unique provider id?). Why this direction and not a map of provider id to label?

Lastly, if you are updating your toolchains, have you seen the design for [Toolchain Transitions](https://github.com/bazelbuild/proposals/blob/master/designs/2019-02-12-toolchain-transitions.md), including the [migration steps](https://github.com/bazelbuild/proposals/blob/master/designs/2020-02-07-toolchain-transition-migration.md)? This will allow you to declare some toolchain dependencies as being in the target config (for the target being built with the toolchain), and others in the exec config, which may be useful with these deps.

Reply:
> I looked over the PR and the docs, and I'm afraid I don't fully understand what's being implemented. You are adding dependencies to the toolchain that will then also be provided to other, non-toolchain-aware rules, via the normal transitive dependencies? Can you give a concrete example of where this would be useful?

Example: [scalac](https://github.com/bazelbuild/rules_scala/blob/master/src/java/io/bazel/rulesscala/scalac/BUILD#L23) target is `java_binary` which isn't aware of `scala_toolchain`.

> 
> As for the map, I am also confused about why it is a map from a label (which seems to always be a declare_deps_provider) to a string (which is a unique provider id?). Why this direction and not a map of provider id to label?

I haven't found a type for string to label mapping (https://docs.bazel.build/versions/master/skylark/lib/attr.html). I don't like having inverted dict as a compromise, but I failed to find a better alternative.

> Lastly, if you are updating your toolchains, have you seen the design for [Toolchain Transitions](https://github.com/bazelbuild/proposals/blob/master/designs/2019-02-12-toolchain-transitions.md), including the [migration steps](https://github.com/bazelbuild/proposals/blob/master/designs/2020-02-07-toolchain-transition-migration.md)? This will allow you to declare some toolchain dependencies as being in the target config (for the target being built with the toolchain), and others in the exec config, which may be useful with these deps.

I've looked at toolchain transitions, but I do not see how it fits into our case. Are there good examples to learn from?

Reply:
If scalac is a java_binary, why doesn't it have a complete set of deps? Can it only be built as a dependency of a scala target? Or have I misunderstood the build graph?

Instead of a string_to_label dict, the typical approach is to use a label_list, and when you get the DepInfo providers (is that the right provider from this attribute?) from the attribute, each DepsInfo has the provider id. Are you ever actually using the same target as different provider ids in different toolchains?

Toolchain transitions are new, and we're still in the migration period. The basic idea is that it allows your toolchain to separate out the tool dependencies (like scalac), which need to run on the execution platform, from your library dependencies (which are actually linked into your final target), and which need to be in the target configuration.

Reply:
I'll move provider id to DepsInfo. I don't like this approach, but it is semantically more correct. Mapping ids to labels feels more declarative and would be good for mapping the same targets more than once, but inverted map does not allow the same label twice anyway.

Reply:
Thanks, I agree that the previous labe->id map was odd. Feel free to reach out to the Starlark API team on bazel-users and ask about this, but I suspect the answer will be that the label_list system you have is more "bazel-y".

In general, I really like this approach for solving the problem of adding transitive deps to things that consume scala targets. I think linking future toolchain authors to your docs will definitely be useful. Would it be possible for the author to also write a blog post or some other article with a high-level discussion to point other rule authors to? I realize this is even more work, but it could really help the community.